typing/typecore.ml

```diff
@@ -670,92 +674,109 @@ end) = struct
         in
         List.find check_type lbls
 
+  (** [disambiguate] selects a concrete description for [lid] using some
+      contextual information: an optional [expected_type], and a list of candidates [lbls].
+      If [expected_type] is [None], it just returns the head of [lbls].
+      Otherwise, it extracts the description directly from the corresponding type definition,
+      except for extension types, where it has to choose it inside [lbls].
+      This code is careful to warn in a number of situations:
+      * if a non principal type was used for disambiguation [Not_principal]
+      * if [lbls] contained several valid candidates [Ambiguous_name]
+      * if the name was chosen out of [scope] [Name_out_of_scope]
+      Here [scope] contains the labels equal to [lid] in the current environment,
+      and [lbls] is a (potentially strict) subset of [scope], see [disambiguate_label_by_ids] *)
```

Comment:
I think that these two remarks could be helpful to have *above* the rest of the specification, otherwise it's hard to follow the explanation if we assume that `lbls` is just an arbitrary list of labels (where are they coming from?).

Reply:
Note: reading this API suggests that the code could be clearer if instead of taking a `scope, lbls` pair whose relation requires explanations, the code of `disambiguate` was parameterized over a filtering function `labels -> (labels, labels) result` that it would call itself on `scope`.

Reply:
`scope` only appears on the previous line, and I'm not sure this information is needed to understand anything else.
The semantics of `disambiguate` does not require this inclusion, but if it is not true the fact we first look up among `lbls` matters (i.e. we could have a solution that wouldn't be reached by the type, even outside of extension types).

We can remove the explanation relating `lbls` and `scope`, and rewrite the explanation above to follow strictly the code:
```
Otherwise we try to find a description in `lbls` with a matching type,
or, if this fails, extract a description from the type definition itself.
```
Then we don't have to refer to extension types at all.

By the way, I forgot the last warning:
```
* if the selected name was not the same as before disambiguation was introduced [Disambiguated_name]
```

Reply:
> We can remove the explanation relating lbls and scope, and rewrite the explanation above to follow strictly the code:
> ```
> Otherwise we try to find a description in `lbls` with a matching type,
> or, if this fails, extract a description from the type definition itself.
> ```
> Then we don't have to refer to extension types at all.

That would be a mistake: if you do that, you're not explaining anything, you're just paraphrasing the code. In this instance I think that the behavior was already obvious from the code (but others might disagree, in which case we can try and clarify *the code*), but I didn't know why that behavior was chosen.
A good *explanation* (mentioning extensible types) would be useful.

Reply:
I agree with @trefis' point that the comment is closer to a description of the implementation (than to a specification). For me the parts that really helps reading the code is the explanation of what `lbls` and `scope` are (and maybe we could avoid need with clearer names, eg. `candidates_in_scope` and `candidates_selection`).

The part on the warnings is somewhat repetitive of the warnings themselves, I will get rid of it.

I'm interested in feedback on the other parts. (Personally I think that having some description is good, it could be shorter, but there is also some confusion, for example calling `lbls` a list when their type is not a list.)

Reply:
OK, let's keep the old version.
And for the last two lines, I just copied what you had written :)
I'm fine with replacing `lbls` by `candidates`.
For `scope`, this is a bit more complicated. The only reason we keep `scope` independently of `lbls` is for cases where type-checking will eventually fail: a record construction with incompatible or missing labels, so that the label is eliminated from `lbls`, and rediscovered through the type. So I would not call its members `candidates_in_scope`, since they  were already discarded, and rather say `candidate_scope` if you want to be more explicit.

For the question about the use of a `return` type, I'm not sure it is essential to understand the code. I suppose it is clear that we are talking about the contents of the `return` type.
Maybe the person who introduced it can add a comment.

Reply:
The result types were introduced by @lpw25 in c19e8b23506e38fc47c110138d50b57fe93f7d7e.

Reply:
Yes, it allows to keep all lookup logic in `Env`, previously it was spread all over the place. You call `lookup_all_foo` and get a `(foo list, lookup_error) result`. Either you have some `foo`s (ideally it would use a non-empty list type, but we don't have a convenient one of those lying around), or you have an error describing why we couldn't find any.

Reply:
The point here, I think, is that the caller passes a `('a, error) result` instead of forcing the `'a` itself, because we don't know yet whether the value will in fact be needed (it is only needed if type-based disambiguation is not possible or is not principal), and we want to error only when necessary.

Reply:
Also, I don't think that removing the part on warnings is a good idea: a function name is not a specification. You might rather argue that my description was not precise enough.

Reply:
> and we want to error only when necessary.

That's right yes. Some of this complication is because the code mixes up the disambiguation code, which can work on long identifiers like `M.Foo`, with the code for type-based lookup, which only works on short identifiers like `Foo`. The long case potentially has interesting errors -- maybe it is `M` that doesn't exist rather than `Foo` -- but should always print the error if it exists. The short case has only a single error -- i.e. there is no `Foo` in scope -- but can ignore the error if it can find the constructor via the type.